Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -788,40 +788,35 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     if shouldSkip(node) {
       return .skipChildren
     }
-    if let unexpected = node.unexpectedBetweenDeinitKeywordAndEffectSpecifiers,
-      let name = unexpected.presentTokens(satisfying: { $0.tokenKind.isIdentifier == true }).only
-    {
-      addDiagnostic(
-        name,
-        .deinitCannotHaveName,
-        fixIts: [
-          FixIt(message: RemoveNodesFixIt(name), changes: .makeMissing(name))
-        ],
-        handledNodes: [name.id]
-      )
-    }
-    if let unexpected = node.unexpectedBetweenDeinitKeywordAndEffectSpecifiers,
-      let params = unexpected.compactMap({ $0.as(FunctionParameterClauseSyntax.self) }).only
-    {
-      addDiagnostic(
-        params,
-        .deinitCannotHaveParameters,
-        fixIts: [
-          FixIt(message: RemoveNodesFixIt(params), changes: .makeMissing(params))
-        ],
-        handledNodes: [params.id]
-      )
-    }
-    if let unexpected = node.unexpectedBetweenEffectSpecifiersAndBody,
-      let returnType = unexpected.compactMap({ $0.as(ReturnClauseSyntax.self) }).only
-    {
+
+    let name: TokenSyntax? = node
+      .unexpectedBetweenDeinitKeywordAndEffectSpecifiers?
+      .presentTokens(satisfying: \.tokenKind.isIdentifier)
+      .only
+
+    let params: FunctionParameterClauseSyntax? = node
+      .unexpectedBetweenDeinitKeywordAndEffectSpecifiers?
+      .compactMap({ $0.as(FunctionParameterClauseSyntax.self) })
+      .only
+
+    let returnType: ReturnClauseSyntax? = node
+      .unexpectedBetweenEffectSpecifiersAndBody?
+      .compactMap({ $0.as(ReturnClauseSyntax.self) })
+      .only
+
+    let nodes = [Syntax(name), Syntax(params), Syntax(returnType)].compactMap { $0 }
+
+    if !nodes.isEmpty {
       addDiagnostic(
-        returnType,
-        .deinitCannotHaveReturnType,
+        node,
```

Comment:
Could we anchor this at `nodes.first!`? That way the diagnostic actually shows up at the error unexpected nodes instead of the `deinit` keyword. And also, I would add a `highlight` for all the unexpected nodes (would be worth checking in at least one test case as well).